lib/util.ts

```diff
@@ -84,24 +84,37 @@ export class BroadcastUtil {
 
   /**
    * Given the indices of a broadcasted tensor, calculate the original indices
-   * @param indices The given indices of the broadcasted tensor.
+   * @param broadcastedIndices The given indices of the broadcasted tensor.
+   * @param originalShape The original shape of the tensor before broadcas
+   * @returns The calculated indices that maps to the original tensor.
+   */
+  static index(broadcastedIndices: ReadonlyArray<number>, originalShape: ReadonlyArray<number>): number[] {
+    // NOTE 1: we assume the parameter broadcastedIndices is valid. ie. it should have the same
+    // length as the broadcasted shape, and for each dimension the index should
+    // not be out of range.
+    const originalIndices = new Array(originalShape.length);
+    BroadcastUtil.fillIndex(broadcastedIndices, originalShape, originalIndices);
+    return originalIndices;
+  }
+
+  /**
+   * Given the indices of a broadcasted tensor, calculate the original indices
+   * @param broadcastedIndices The given indices of the broadcasted tensor.
    * @param originalShape The original shape of the tensor before broadcast
-   * @param isMatMul Whether the operation is MatMul
-   * @returns The calculated indices that maps to the original tensor. If the
-   * operation is MatMul, the indices of last 2 dimensions will keep as same as
-   * input indices
+   * @param originalIndices The mapping of broadcastedIndices to the originalIndices (output parameter - will be
+   *     mutated).
    */
-  static index(indices: ReadonlyArray<number>, originalShape: ReadonlyArray<number>, isMatMul = false): number[] {
-    // we assume the parameter indices is valid. ie. it should have the same
+  static fillIndex(
+      broadcastedIndices: ReadonlyArray<number>, originalShape: ReadonlyArray<number>, originalIndices: number[]) {
+    // NOTE 1: we assume the parameter broadcastedIndices is valid. ie. it should have the same
     // length as the broadcasted shape, and for each dimension the index should
     // not be out of range.
-    const dimOffset = indices.length - originalShape.length;
-    const indicesOriginal = indices.slice(dimOffset);
-    const dimLen = isMatMul ? indicesOriginal.length - 2 : indicesOriginal.length;
-    for (let i = 0; i < dimLen; i++) {
-      indicesOriginal[i] = indices[dimOffset + i] % originalShape[i];
+    // NOTE 2: we assume the parameter originalIndices has the same
```

Comment:
comments can be put in oneline

Reply:
For NOTE 2, I made it single lined. For NOTE 1, it cannot be made a one-liner and I would prefer to keep both separate as they have differing themes.